server.go

```diff
@@ -116,9 +161,15 @@ func (server *Server) Stop() (err error) {
 }
 
 // Public interface to get access to server messages
-// Returns slice of message pointers
-func (server *Server) Messages() []*Message {
-	return server.messages.items
+// Returns copy of message data
+func (server *Server) Messages() []Message {
```

Comment:
Does this implementation not covers thread safe using for read/write operations?

https://github.com/mocktools/go-smtp-mock/blob/03e6fbe72340d8081c7ab1ed091c7de18c93367c/message.go#L126

```go
// Concurrent type that can be safely shared between goroutines
type messages struct {
	sync.RWMutex
	items []*Message
}

// Addes new message pointer into concurrent messages slice
func (messages *messages) append(item *Message) {
	messages.Lock()
	defer messages.Unlock()

	messages.items = append(messages.items, item)
}
```

Reply:
Yes, your use of `append` is correct, the problem is that you aren't factoring that you still have a race condition on the READ side when you return a slice of pointers to exactly the same instances of `Message`. There's no use of a mutex on `*Message` is there? You have to worry about concurrent read as much as concurrent write. This PR only addresses the lack of thread safety in the `Server` type, but, yes, I would say the getter methods on `*Message` type (e.g. `MailfromRequest()` need hardening as well. I think this can be done simply by changing the type that the methods are on from `*Message` to `Message`.

Moreover, I'm saying that it's not safe from a _data_ perspective to return pointers to shared data. Furthermore, there is absolutely no need for it. You shouldn't ever return a pointer to something unless for a very specific reason. For the public interface to get a view of the messages at a point in time? There is no reason, and for this purpose the user is expecting a _copy_ of the data.

Reply:
@benjamin-rood 

> Moreover, I'm saying that it's not safe from a data perspective to return pointers to shared data, ever. Furthermore, there is absolutely no need for it. You shouldn't ever return a pointer to something unless for a very specific reason. For the public interface to get a view of the messages at a point in time? There is no reason, and for this purpose the user is expecting a copy of the data.

My goal was to avoid additional resources allocation when it is not necessary and avoid the Ruby way `in any case just return the copy of object`. Maybe make sense to implement thread safety reading of this messages instead of returning copies?

Reply:
> My goal was to avoid additional resources allocation when it is not necessary and avoid the Ruby way `in any case just return the copy of object`. Maybe make sense to implement thread safety reading of this messages instead of returning copies?

I understand the intent, but it's an unnecessary optimisation. A little copying is better than a little problem, to paraphrase Rob Pike. Never do something until there is a proven need. Moreover, remember that copying is much, much faster in compiled Go than Ruby, so, this is really not something to worry about; once we are expecting to copy thousands of things every few milliseconds in a real application context then it's worth some consideration, and, even then it wouldn't be the correct approach in this case. Copying is safe, copying is correct.

Reply:
@benjamin-rood Thanks, finally I got your point of view. I agree 👍